torchrl/envs/transforms/transforms.py

```diff
@@ -2429,3 +2429,50 @@ def __repr__(self) -> str:
             f"{self.__class__.__name__}(decay={self.decay:4.4f},"
             f"eps={self.eps:4.4f}, keys={self.in_keys})"
         )
+
+
+class StepCounter(Transform):
+    """Counts the steps from a reset and sets the done state to True after a certain number of steps.
+
+    Args:
+        max_steps (:obj:`int`, optional): the maximum number of steps to take before setting the done state to True.
+    """
+
+    invertible = False
+    inplace = True
+
+    def __init__(self, max_steps: Optional[int]):
+        self.max_steps = max_steps
+        super().__init__([])
+
+    def reset(self, tensordict: TensorDictBase) -> TensorDictBase:
+        tensordict.set(
+            "step_count", torch.zeros(*tensordict.batch_size, 1, dtype=torch.int64)
+        )
+        if self.max_steps is not None and self.max_steps <= 0:
+            tensordict.fill_("done", True)
+        return tensordict
+
+    def _step(self, tensordict: TensorDictBase) -> TensorDictBase:
+        next_step_count = tensordict.get("step_count") + 1
+        tensordict.set("step_count", next_step_count)
+        if self.max_steps is not None:
+            tensordict.set(
+                "done",
+                torch.where(
+                    next_step_count < self.max_steps, tensordict.get("done"), True
+                ),
+            )
+        return tensordict
+
+    def _transform_spec(self, spec: TensorSpec) -> None:
+        if isinstance(spec, CompositeSpec):
+            for key in spec:
+                self._transform_spec(spec[key])
+        else:
+            spec.dtype = torch.int64
+
+    @_apply_to_composite
+    def transform_observation_spec(self, observation_spec: TensorSpec) -> TensorSpec:
+        self._transform_spec(observation_spec)
```

Comment:
Maybe i'm missing something, but I think you should do something like
```
    def transform_observation_spec(self, observation_spec: TensorSpec) -> TensorSpec:
        observation_spec["step_count"] = UnboundedDiscreteTensorSpec(dtype=torch.int64)
```

Reply:
Thanks for all your suggestions! I made the changes, but I would like to know how can I test the `transform_observation_spec` method

Reply:
There's a test function you can run
```
from torchrl.envs.utils import check_env_specs
env = TransformedEnv(GymEnv("Pendulum-v1"), StepCounter(50))
check_env_specs(env)
```

It will run a small rollout + get fake data from your specs, and check that they match.

Reply:
I get this error, I don't really know what I missed,  it comes from the shape of the tensor is []:

```python
test/test_transforms.py:1600 (TestTransforms.test_step_counter_observation_spec)
Traceback (most recent call last):
  File "/Users/waris/Projects/rl/test/test_transforms.py", line 1604, in test_step_counter_observation_spec
    check_env_specs(env)
  File "/Users/waris/Projects/rl/torchrl/envs/utils.py", line 168, in check_env_specs
    fake_tensordict = env.fake_tensordict().flatten_keys(".")
  File "/Users/waris/Projects/rl/torchrl/envs/common.py", line 636, in fake_tensordict
    fake_obs = observation_spec.rand(self.batch_size)
  File "/Users/waris/Projects/rl/torchrl/data/tensor_specs.py", line 1286, in rand
    _dict = {
  File "/Users/waris/Projects/rl/torchrl/data/tensor_specs.py", line 1287, in <dictcomp>
    key: self[key].rand(shape)
  File "/Users/waris/Projects/rl/torchrl/data/tensor_specs.py", line 655, in rand
    r = torch.rand(*shape, *interval.shape, device=interval.device)
TypeError: rand() missing 1 required positional arguments: "size"
```

Reply:
On line 655, try putting the shape in between brackets
Like `rand([*shape1, *shape2],...)`
Also make sure you've merged main into your branch!